README
NAME

    spaceless - Remove spaces and other hazards from your PATH

VERSION

    version 0.06

SYNOPSIS

    Convert PATH (by default):

     cygwin% spaceless
     PATH=/usr/bin:/cygdrive/c/PROGRA~2/NVIDIA~1/PhysX/Common:...
     export PATH

    Convert another PATH style environment variable

     cygwin% spaceless PERL5LIB
     PERL5LIB=/PERL~1/lib:/PERL~2/lib
     export PERL5LIB

    Update the PATH in the current shell (works with both sh and csh):

     cygwin% eval `spaceless PATH`

    Same thing from cmd.exe or command.com prompt:

     C:\> spaceless PATH -f path.bat
     C:\> path.bat

DESCRIPTION

    Although legal in most modern operating system directory names, spaces
    are only frequently encountered in Windows. This can be troublesome for
    some tools that do not take such antics into account. Fortunately,
    Windows provides alternate file and directory names for any name that
    is not compatible with MS DOS. Since spaces were never supported by MS
    DOS we can use these alternate names for tools that don't take spaces
    into account.

    spaceless converts PATH style environment variables on windows into
    equivalents that do not have spaces. By default it uses Shell::Guess to
    make a reasonable guess as to the shell you are currently using (not
    your login shell). You may alternately specify a specific shell type
    using one of the options below.

    spaceless will not convert long directory (non 8.3 format) names that
    do not have spaces, as these are usually handled gracefully by scripts
    that are space challenged.

OPTIONS

 --cmd

    Generate cmd.exe configuration

 --command

    Generate command.com configuration

 --csh

    Generate c shell configuration

 -f filename

    Write configuration to filename instead of standard output

 --expand | -x

    Expand short paths to long paths (adding any spaces that may be back
    in)

 --fish

    Generate fish shell configuration

 --help

    Print help message and exit

 --korn

    Generate korn shell configuration

 --list | -l

    Instead of generating a shell configuration, list the directories in
    the given path style variable

 --login

    Use the your default login shell (as determined by Shell::Guess) On
    Unix style systems this consults the GECOS field in the /etc/passwd
    file or uses NIS. On other platforms it may use another means to
    determine your login shell, or simply make an informed guess based on
    the platform.

 --no-cygwin

    Remove any cygwin paths. Has no affect on non cygwin platforms.

 --power

    Generate Power shell configuration

 --sh

    Generate bourne shell configuration

 --sep char | --sep-in char | --sep-out char

    You can use these options to change the path separator character.
    Normally you do not need to set this. the Platform default is used for
    input (: for Unix and cygwin, and ; for Windows) and the shell default
    is used for output (: for sh and csh, and ; for command.com and
    cmd.exe). But in some situation you might want a UNIX style path in a
    cmd.exe script. Any character may be specified, but keep in mind that
    not all characters are supported by Shell::Config::Generate.

 --squash | -s

    Remove duplicate paths from the given path style variable.

 --trim | -t

    Trim non-existing directories. This is a good idea since such
    directories can sometimes cause spaceless to die.

 --version | v

    Print version number and exit

EXAMPLES

 Removing spaces from the PATH on Cygwin

    My motivation for writing this script was trying to get perlbrew to
    work on Cygwin. Since Windows frequently includes spaces in its %PATH%
    environment variable, and cygwin inherits them.

     xian-x86_64% source ~/perl5/perlbrew/etc/cshrc
     setenv: Too many arguments.
     xian-x86_64% eval `spaceless PATH`
     xian-x86_64% source ~/perl5/perlbrew/etc/cshrc
     xian-x86_64%

    I could have manually updated my %PATH% to not include spaces, or
    better yet submitted a patch to perlbrew to fix its spacing problem.
    This probably won't be the last script that I will have the spaces in
    the %PATH% problem with.

 Visualizing the PATH

    Reading the %PATH% variable can be difficult, especially if you have
    already removed the spaces and are using the short version.

     C:\> echo %PATH%
     N:\lang\perl\strawberry\x86\5.18.2\c\bin;N:\lang\perl\strawberry
     \x86\5.18.2\perl\site\bin;N:\lang\perl\strawberry\x86\5.18.2\per
     l\bin;n:\program32\GnuWin32\bin;C:\PROGRA~2\NVIDIA~1\PhysX\Commo
     n;C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Win
     dows\System32\WindowsPowerShell\v1.0\;C:\PROGRA~1\MICROS~3\110\T
     ools\Binn\;C:\PROGRA~2\Git\cmd;N:\lang\tcl\x86\8.6.1.0\bin;C:\PR
     OGRA~2\NVIDIA~1\PhysX\Common;C:\PROGRA~2\Intel\ICLSCL~1\;C:\PROG
     RA~1\Intel\ICLSCL~1\;C:\Windows\system32;C:\Windows;C:\Windows\S
     ystem32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;

    You can use the -l (short for --list) and -x (short for --expand)
    options to make it a little more readable

     C:\> spaceless -l -x
     N:\lang\perl\strawberry\x86\5.18.2\c\bin
     N:\lang\perl\strawberry\x86\5.18.2\perl\site\bin
     N:\lang\perl\strawberry\x86\5.18.2\perl\bin
     N:\program32\GnuWin32\bin
     C:\Program Files (x86)\NVIDIA Corporation\PhysX\Common
     ...

 Remove duplicate directories in the path

    You can use the -s (short for --squeeze) option to squeeze out
    duplicate paths. It keeps the first instance of a directory in a path
    style variable, which is usually what you want.

     % spaceless -l
     /usr/local/bin
     /usr/bin
     /bin
     /usr/local/bin
     % spaceless -s -l
     /usr/local/bin
     /usr/bin
     /bin

    You can then update the PATH with the eval trick:

     % eval `spaceless -s`

AUTHOR

    Graham Ollis <plicease@cpan.org>

COPYRIGHT AND LICENSE

    This software is copyright (c) 2014 by Graham Ollis.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for App-spaceless

0.06      2014-12-20 05:39:56 -0500
  - internal cleanup
  - now requires Perl 5.10.1

0.05      2014-05-31 12:42:32 -0400
  - capture warning from running shell during test
    to avoid failure

0.04      2014-05-31 06:37:09 -0400
  - bump version requirement for Shell::Guess to 0.04
    to avoid infrequent test failure

0.03      2014-05-29 07:43:32 -0400
  - first CPAN release
  - added --login option
  - added --list option
  - added --squeeze option
  - added --sep* options

0.02      2014-05-28 08:53:13 -0400
  - added --expand | -x option
  - added --no-cygwin option
  - added t/basic.t
  - avoid warnings

0.01      2014-05-27 20:33:28 -0400
  - initial version


Build.PL
BEGIN {
  unless(eval q{ use 5.010001; 1}) {
    print "Perl 5.010001 or better required\n";
    exit;
  }
}

# This file was automatically generated by Dist::Zilla::Plugin::ModuleBuild v5.029.
use strict;
use warnings;

use Module::Build 0.28;


my %module_build_args = (
  "build_requires" => {
    "Module::Build" => "0.28"
  },
  "configure_requires" => {
    "Module::Build" => "0.28",
    "perl" => "5.010001"
  },
  "dist_abstract" => "Convert PATH type environment variables to spaceless versions",
  "dist_author" => [
    "Graham Ollis <plicease\@cpan.org>"
  ],
  "dist_name" => "App-spaceless",
  "dist_version" => "0.06",
  "license" => "perl",
  "module_name" => "App::spaceless",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Shell::Config::Generate" => "0.17",
    "Shell::Guess" => "0.04",
    "perl" => "5.010001"
  },
  "script_files" => [
    "bin/spaceless"
  ],
  "test_requires" => {
    "Capture::Tiny" => 0,
    "Path::Class" => "0.26",
    "Test::More" => "0.94",
    "perl" => "5.010001"
  }
);


my %fallback_build_requires = (
  "Capture::Tiny" => 0,
  "Module::Build" => "0.28",
  "Path::Class" => "0.26",
  "Test::More" => "0.94",
  "perl" => "5.010001"
);


unless ( eval { Module::Build->VERSION(0.4004) } ) {
  delete $module_build_args{test_requires};
  $module_build_args{build_requires} = \%fallback_build_requires;
}

my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/App/spaceless.pm
package App::spaceless;

use strict;
use warnings;
use 5.010001;
use Config;
use Shell::Guess;
use Shell::Config::Generate qw( win32_space_be_gone );
use Getopt::Long qw( GetOptions );
use Pod::Usage qw( pod2usage );

# ABSTRACT: Convert PATH type environment variables to spaceless versions
our $VERSION = '0.06'; # VERSION


sub _running_shell
{
  state $shell;
  $shell = Shell::Guess->running_shell unless defined $shell;
  $shell;
}

sub main
{
  shift;
  local @ARGV = @_;
  my $shell;
  my $file;
  my $help;
  my $version;
  my $trim;
  my $cygwin = 1;
  my $expand;
  my $list;
  my $squash;

  my $config = Shell::Config::Generate->new;
  $config->echo_off;
  my $sep = quotemeta $Config{path_sep};
  
  GetOptions(
    'csh'       => sub { $shell = Shell::Guess->c_shell },
    'sh'        => sub { $shell = Shell::Guess->bourne_shell },
    'cmd'       => sub { $shell = Shell::Guess->cmd_shell },
    'command'   => sub { $shell = Shell::Guess->command_shell },
    'fish'      => sub { $shell = Shell::Guess->fish_shell },
    'korn'      => sub { $shell = Shell::Guess->korn_shell },
    'power'     => sub { $shell = Shell::Guess->power_shell },
    'login'     => sub { $shell = Shell::Guess->login_shell },
    'sep=s'     => sub { $sep = quotemeta $_[1]; $config->set_path_sep($_[1]) },
    'sep-in=s'  => sub { $sep = quotemeta $_[1] },
    'sep-out=s' => sub { $config->set_path_sep($_[1]) },
    'squash|s'  => \$squash,
    'no-cygwin' => sub { $cygwin = 0 if $^O eq 'cygwin' },
    'list|l'    => \$list,
    'expand|x'  => \$expand,
    'trim|t'    => \$trim,
    'f=s'       => \$file,
    'help|h'    => \$help,
    'version|v' => \$version,
  ) || pod2usage(1);

  if($help)
  {
    pod2usage({ -verbose => 2 });
  }
  
  if($version)
  {
    say 'App::spaceless version ', ($App::spaceless::VERSION // 'dev');
    return 1;
  }

  $shell = _running_shell() unless defined $shell;
  
  my $filter = $^O eq 'cygwin' && $shell->is_win32 ? sub { map { Cygwin::posix_to_win_path($_) } @_ } : sub { @_ };

  @ARGV = ('PATH') unless @ARGV;

  my $to_long = $^O eq 'cygwin' ? sub { Cygwin::win_to_posix_path(Win32::GetLongPathName(Cygwin::posix_to_win_path($_))) } : sub { Win32::GetLongPathName($_[0]) };

  my $mutator = $expand ? sub { map { $to_long->($_) } @_ } : sub { win32_space_be_gone(@_) };

  foreach my $var (@ARGV)
  {
    my @path = $filter->($mutator->(
      grep { $trim ? -d $_ : 1 } 
      grep { $cygwin ? 1 : $_ =~ qr{^([A-Za-z]:|/cygdrive/[A-Za-z])} } 
      split /$sep/, $ENV{$var} // ''
    ));
    
    if($squash)
    {
      my %path;
      @path = grep { !$path{$_}++ } @path;
    }
    
    $config->set_path( $var => @path );
    do { say $_ for @path } if $list;
  }

  unless($list)
  {
    if(defined $file)
    {
      $config->generate_file($shell, $file);
    }
    else
    {
      print $config->generate($shell);
    }
  }
  
  return 0;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

App::spaceless - Convert PATH type environment variables to spaceless versions

=head1 VERSION

version 0.06

=head1 DESCRIPTION

This module provides the machinery for the L<spaceless> app, a program
that helps convert PATH style environment variables to spaceless varieties
on Windows systems (including Cygwin).

=head1 AUTHOR

Graham Ollis <plicease@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Graham Ollis.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut
